### PR TITLE
Adds F.R.A.M.E. cartridge

### PR DESCRIPTION
**Pull request by lordpidey**
:cl: lordpidey
add: Added F.R.A.M.E. cartridge to uplinks.  This PDA cartridge contains 5 viruses, which when used will unlock the target's PDA into a syndicate uplink, in addition, you will receive the code needed to unlock the uplink again, if you so desire.
add: If you use telecrystals on a F.R.A.M.E. cartridge, the next time it is used, it will also give those crystals to the target uplink.
/:cl:

Two things in this PR.

One: I refactored virus based PDA uplinks, they're more consistent now.

Two: I added the F.R.A.M.E. cartridge, it works similarly to the existing virus cartirdges.  It has five charges, and upon using a charge, it will unlock the target's PDA with zero TC.  You will also receive the code needed to unlock the PDA.

If it's used on an already traitorous PDA, it will still unlock, but with zero TC.  Though re-locking and unlocking normally will bring the normal TC count back. (So you can't use it to steal TC from other traitors).  If this happens, the code you receive for unlocking it will not work.

You may also apply TC directly to the cartridge, and the next time it is used, it will not only unlock the PDA, but also give the crystals to the target PDA.

If you use TC on it AND target an already traitorous PDA, it will only show the crystals you put in, but if locked and unlocked, it will simply add the new crystals to the old crystals.

Why:  I think this will allow a great deal of stealth espionage to the game, as it allows stealth transferring of crystals, deputizing sub traitors, and framing people as a traitor.  The possibilities are endless.

### CODE REVIEW

**Comment by kevinz000** [comment]
OH BOY

**Comment by Thunder12345** [comment]
Breaking the meta, one hard tell at a time

**Comment by Cruix** [comment]
I disagree with the idea that PDA cartridge subtypes should have any unique code, because that makes it functionally impossible for you to make one cartridge that has the features of two uniquely-coded subtypes. For example, with this implementation, it is impossible to have a single cartridge with both the mime and clown viruses on it. This is still better than the old virus code, which had the same problems.

All of the virus procs should be at the /obj/item/weapon/cartridge level, and the ability to call them should be set by a variable like all the other PDA cartridge functions.

**Comment by lordpidey** [comment]
I can think of one alternative method, but it would be ugly as hell, keeping it both object oriented AND extendable.

I'll try to implement it tomorrow, but I'm not sure it'll be an improvement.

**Comment by octareenroon91** [comment]
What do you want, datum viruses?

**Comment by Fox-McCloud** [comment]
I do recommend making it so that if it forces open the PDA of someone who has TC, it displays the TC they have, otherwise, it's going to get confusing.

Likewise if you sent them TC; it should unlock it and display the *new* total, not how much was sent....again, to avoid confusion.

**Comment by Bawhoppen** [comment]
this would just make a new meta

**Comment by octareenroon91** [comment]
>Or is the reason this exists purely to avoid sec popping open a suspected antag's uplink and confirming it by the difference in TC?

Yes, that and to prevent traitors from gaining anything from stealing another's PDA.

**Comment by lordpidey** [comment]
>Or is the reason this exists purely to avoid sec popping open a suspected antag's uplink and confirming it by the difference in TC?

Well, there would be one other psuedo tell availible even with the TC difference.  It's more to prevent TC theft.

Take the target's PDA, apply the virus, close the uplink, attempt to re-open the uplink with the new code.  If it does NOT open, then either the target was previously hit by the virus, or they are a traitor.

File: code/game/objects/items/devices/PDA/virus_cart.dm [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	if(!isnull(target) && !target.toff)
+		charges--
+		var/lock_code = "[rand(100,999)] [pick("Alpha","Bravo","Charlie","Delta","Echo","Foxtrot","Golf","Hotel","India","Juliet","Kilo","Lima","Mike","November","Oscar","Papa","Quebec","Romeo","Sierra","Tango","Uniform","Victor","Whiskey","X-ray","Yankee","Zulu")]"
+		to_chat(U, "<span class='notice'>Virus Sent!  The unlock code to the target is: [lock_code]</span>")
+		if(!target.hidden_uplink)
+			var/obj/item/device/uplink/uplink = new(target)
+			target.hidden_uplink = uplink
+			target.lock_code = lock_code
+		else
+			target.hidden_uplink.hidden_crystals = target.hidden_uplink.telecrystals //Temporarially hide the PDA's crystals, so you can't steal telecrystals.
```
Comment (AnturK) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: Won't old hidden tc be lost if framed twice without closing?

Comment (lordpidey) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: Oh jeez.  That's a great point.  I'll get this one later today.

**Pull request merged**